divide_and_conquer/Strassen’s Matrix Multiplication.cpp

```diff
@@ -0,0 +1,182 @@
+	else {
+		int split_index = col_1 / 2;
+
+		vector<int> row_vector(split_index, 0);
+
+		vector<vector<int> > a00(split_index, row_vector);
+		vector<vector<int> > a01(split_index, row_vector);
+		vector<vector<int> > a10(split_index, row_vector);
+		vector<vector<int> > a11(split_index, row_vector);
+		vector<vector<int> > b00(split_index, row_vector);
+		vector<vector<int> > b01(split_index, row_vector);
+		vector<vector<int> > b10(split_index, row_vector);
+		vector<vector<int> > b11(split_index, row_vector);
+
+		for (auto i = 0; i < split_index; i++)
+			for (auto j = 0; j < split_index; j++) {
+				a00[i][j] = matrix_A[i][j];
+				a01[i][j] = matrix_A[i][j + split_index];
+				a10[i][j] = matrix_A[split_index + i][j];
+				a11[i][j] = matrix_A[i + split_index]
+									[j + split_index];
+				b00[i][j] = matrix_B[i][j];
+				b01[i][j] = matrix_B[i][j + split_index];
+				b10[i][j] = matrix_B[split_index + i][j];
+				b11[i][j] = matrix_B[i + split_index]
+									[j + split_index];
+			}
+
+		vector<vector<int> > p(multiply_matrix(
+			a00, add_matrix(b01, b11, split_index, -1)));
+		vector<vector<int> > q(multiply_matrix(
+			add_matrix(a00, a01, split_index), b11));
+		vector<vector<int> > r(multiply_matrix(
+			add_matrix(a10, a11, split_index), b00));
+		vector<vector<int> > s(multiply_matrix(
+			a11, add_matrix(b10, b00, split_index, -1)));
+		vector<vector<int> > t(multiply_matrix(
+			add_matrix(a00, a11, split_index),
+			add_matrix(b00, b11, split_index)));
+		vector<vector<int> > u(multiply_matrix(
+			add_matrix(a01, a11, split_index, -1),
+			add_matrix(b10, b11, split_index)));
+		vector<vector<int> > v(multiply_matrix(
+			add_matrix(a00, a10, split_index, -1),
+			add_matrix(b00, b01, split_index)));
+
+		vector<vector<int> > result_matrix_00(add_matrix(
+			add_matrix(add_matrix(t, s, split_index), u,
+					split_index),
+			q, split_index, -1));
+		vector<vector<int> > result_matrix_01(
+			add_matrix(p, q, split_index));
+		vector<vector<int> > result_matrix_10(
+			add_matrix(r, s, split_index));
+		vector<vector<int> > result_matrix_11(add_matrix(
+			add_matrix(add_matrix(t, p, split_index), r,
+					split_index, -1),
+			v, split_index, -1));
+
+		for (auto i = 0; i < split_index; i++)
+			for (auto j = 0; j < split_index; j++) {
+				result_matrix[i][j]
+					= result_matrix_00[i][j];
+				result_matrix[i][j + split_index]
+					= result_matrix_01[i][j];
+				result_matrix[split_index + i][j]
+					= result_matrix_10[i][j];
+				result_matrix[i + split_index]
+							[j + split_index]
+					= result_matrix_11[i][j];
+			}
+
+		a00.clear();
+		a01.clear();
+		a10.clear();
+		a11.clear();
+		b00.clear();
+		b01.clear();
+		b10.clear();
+		b11.clear();
+		p.clear();
+		q.clear();
+		r.clear();
+		s.clear();
+		t.clear();
+		u.clear();
+		v.clear();
+		result_matrix_00.clear();
+		result_matrix_01.clear();
+		result_matrix_10.clear();
+		result_matrix_11.clear();
+	}
+	return result_matrix;
+}
+
+int main()
+{
```

Comment:
There is no test for correctness of the multiplication.

Reply:
...a nice test implementation is here: https://martin-thoma.com/matrix-multiplication-python-java-cpp/#The_Tests